csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -57,58 +57,76 @@ class CircularBufferLoopCloner : public kir::IrVisitor {
       : circular_buffer_loop_(circular_buffer_loop),
         circular_buffer_load_exprs_(circular_buffer_load_exprs),
         loop_type_(loop_type),
-        exclude_(exclude) {}
+        exclude_(exclude) {
+    std::transform(
+        circular_buffer_load_exprs_.begin(),
+        circular_buffer_load_exprs_.end(),
+        std::inserter(
+            circular_buffer_load_tvs_, circular_buffer_load_tvs_.begin()),
+        [](Expr* load_expr) { return ir_utils::getTvOutput(load_expr); });
+  }
 
   using kir::IrVisitor::handle;
 
-  void clone() {
-    const auto gpu_lower = GpuLower::current();
-
+  void duplicate() {
     // Cloning the circular buffer loop as follows:
     //
     // Prologue: 0 to 1
     // Main: 0 to (extent-1)
     // Epilogue: (extent-1) to extent
 
-    auto index = GpuLower::current()->caMap()->getIndexVariable(
+    Val* index = GpuLower::current()->caMap()->getIndexVariable(
         circular_buffer_loop_->iter_domain(), loop_type_);
-    auto start = circular_buffer_loop_->start();
-    auto stop = circular_buffer_loop_->stop();
-    auto stage_depth = gpu_lower->circularBufferInfo().getStageDepthFor(
-        circular_buffer_loop_->iter_domain());
-
-    if (loop_type_ == CircularBufferLoopStage::Prolog) {
-      NVF_ERROR(start->isZeroInt());
-      stop = SimplifyingIrBuilder::create<Val>(
-          int64_t(stage_depth - 1), DataType::Index);
-    } else if (
-        loop_type_ == CircularBufferLoopStage::Main &&
-        requireEpilogue(circular_buffer_load_exprs_)) {
-      stop = IrBuilder::subExpr(
-          circular_buffer_loop_->stop(),
-          SimplifyingIrBuilder::create<Val>(stage_depth - 1, DataType::Index));
-    } else if (loop_type_ == CircularBufferLoopStage::Epilog) {
-      NVF_ERROR(requireEpilogue(circular_buffer_load_exprs_));
-      start = IrBuilder::subExpr(
-          circular_buffer_loop_->stop(),
-          SimplifyingIrBuilder::create<Val>(stage_depth - 1, DataType::Index));
+    Val* start = circular_buffer_loop_->start();
+    Val* stop = circular_buffer_loop_->stop();
+    int64_t stage_depth =
+        GpuLower::current()->circularBufferInfo().getStageDepthFor(
+            circular_buffer_loop_->iter_domain());
+
+    switch (loop_type_) {
+      case CircularBufferLoopStage::Prolog: {
+        NVF_ERROR(start->isZeroInt());
+        stop = SimplifyingIrBuilder::create<Val>(
+            int64_t(stage_depth - 1), DataType::Index);
+        break;
+      }
+      case CircularBufferLoopStage::Main: {
+        if (requireEpilogue(circular_buffer_load_exprs_)) {
+          stop = IrBuilder::subExpr(
+              circular_buffer_loop_->stop(),
+              SimplifyingIrBuilder::create<Val>(
+                  stage_depth - 1, DataType::Index));
+        }
+        break;
+      }
+      case CircularBufferLoopStage::Epilog: {
+        NVF_ERROR(requireEpilogue(circular_buffer_load_exprs_));
+        start = IrBuilder::subExpr(
```

Comment:
Off-topic a bit, but should these `IrBuilder` calls be `SimplifyingIrBuilder`?

Reply:
It probably doesn't hurt, but I think the scalar hoist pass does most of the simplification.

Reply:
I updated it `SimplifyingIrBuilder` for uniformity with the other case blocks.

Reply:
I reverted the change. We can generate the epilogue for-loop unnecessarily when using `SimplifyingIrBuilder`.
See the failure for test `LoopRotationTest.MultipleCircularBuffer` in [ee6f824](https://github.com/NVIDIA/Fuser/pull/2747/commits/ee6f824b27df81aa4c9dfb1a0b90c6bea7865bb7)  for details.